### PR TITLE
feature/285/WifiCoverage

### PR DESCRIPTION
**Pull request by t2m4k1**
Implemented Android WifiScans for generating a wifi-heatmap.
This implementation was tested for a few days locally and proved stable.
It is necessary to merge the backend changes as well for making it work.
The basic idea is that every TCA user generates WifiMeasurements in 
random intervals. At home I've set the pause times to alternate inbetween
the range 10s to 2min and the battery consumption measured by my samsung
power monitor improved only slightly. In the current commit the timespan is set
about 3 times as big (until 7min) and therefore I expect almost no raise in power
consumption at all. Currently the accumulated scans are sent to the backend with
every iteration of the StartSyncReceiver, which is set to 3hours.
I've let the scans run for a few days and they also continued when the screen was turned off / app was closed. Some data generated for my local wifi (including my address :P) can be seen in the screenshot.

database

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -606,12 +606,18 @@
         <receiver android:name=".services.StartSyncReceiver">
             <intent-filter>
                 <action android:name="android.net.wifi.WIFI_STATE_CHANGED" />
-                <action android:name="android.net.wifi.SCAN_RESULTS" />
```
Comment (kordianbruck) [inline app/src/main/AndroidManifest.xml]: This is actually used for the eduroam setup or not? Why did you remove it?

File: app/src/main/java/de/tum/in/tumcampusapp/api/TUMCabeClient.java [hunk 2]
```diff
@@ -56,6 +57,8 @@
     private static final String API_OWN_QUESTIONS = "question/my/";
     private static final String API_FACULTY = "faculty/";
 
+    private static final String API_WIFI_HEATMAP = "wifimap/";
+
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/api/TUMCabeClient.java]: Please fix the blank lines here. Move the thing one up and remove that added blank line

File: app/src/main/java/de/tum/in/tumcampusapp/api/TUMCabeClient.java [hunk 3]
```diff
@@ -216,6 +219,10 @@ public void deviceUploadGcmToken(DeviceUploadGcmToken verification, Callback<TUM
         service.deviceUploadGcmToken(verification).enqueue(cb);
     }
 
+    public void createMeasurement(WifiMeasurement wifiMeasurement) throws IOException{
+        service.createMeasurement(wifiMeasurement).execute().body();
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/api/TUMCabeClient.java]: Please make this an async function using `enqueue`:
https://futurestud.io/tutorials/retrofit-synchronous-and-asynchronous-requests
https://medium.com/@shelajev/how-to-make-http-calls-on-android-with-retrofit-2-cfc4a67c6254

File: app/src/main/java/de/tum/in/tumcampusapp/auxiliary/Utils.java [hunk 4]
```diff
@@ -555,4 +558,14 @@ public static Spanned fromHtml(String source) {
                 Html.fromHtml(source, Html.FROM_HTML_MODE_LEGACY) :
                 Html.fromHtml(source);
     }
+
+    public static float getBatteryLevel(Context context) {
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/auxiliary/Utils.java]: Was this just for your testing purposes or is this needed in the future?

File: app/src/main/java/de/tum/in/tumcampusapp/auxiliary/WifiMeasurementLocationListener.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+ * which later gets synced with the remote database to generate the wifi heatmap. Only if the measurement
+ * is not older than MAX_MILLISECONDS_PASSED when the location is found it will get stored. Additionally
+ * the measurement location data needs to fulfill the MINIMUM_ACCURACY_IN_METERS.
+ */
+
+public class WifiMeasurementLocationListener implements LocationListener{
+    private Context context;
+    private WifiMeasurement wifiMeasurement;
+    private static final int MAX_MILLISECONDS_PASSED = 3000;
+    private static final int MINIMUM_ACCURACY_IN_METERS = 25;
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/auxiliary/WifiMeasurementLocationListener.java]: Please add comments what these `final`s are

File: app/src/main/java/de/tum/in/tumcampusapp/auxiliary/WifiMeasurementLocationListener.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        this.measurementTakenInMillis = measurementTakenInMillis;
+    }
+
+    /**
+     * This method modifies the measurement it's responsible for by adding location data to it.
+     * If it fulfills the minimum requirements to be stored, it is then saved to the local database.
+     * @param location
+     */
+        @Override
+        public void onLocationChanged(Location location) {
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/auxiliary/WifiMeasurementLocationListener.java]: Why is this suddenly indented with more spaces?

File: app/src/main/java/de/tum/in/tumcampusapp/auxiliary/WifiMeasurementLocationListener.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+            String date = Utils.getDateString(Calendar.getInstance().getTime());
+            WifiMeasurementManager wifiMeasurementManager = new WifiMeasurementManager(context);
+            if (Calendar.getInstance().getTimeInMillis() - measurementTakenInMillis > MAX_MILLISECONDS_PASSED) {
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/auxiliary/WifiMeasurementLocationListener.java]: How does this have anything to do with location? Please add comments in the code, so I can understand if everything is alright.

File: app/src/main/java/de/tum/in/tumcampusapp/auxiliary/WifiMeasurementLocationListener.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+                Utils.log("Wifi Measurement thrown away, location stale!");
+                return;
+            }
+            if (location.getAccuracy() < MINIMUM_ACCURACY_IN_METERS){
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/auxiliary/WifiMeasurementLocationListener.java]: You add this condition to the previous and thus make it more readable (invert it possibly)

File: app/src/main/java/de/tum/in/tumcampusapp/managers/WifiMeasurementManager.java [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+        String SSID = atPosition.getString(2);
+        String BSSID = atPosition.getString(3);
+        int dBm = atPosition.getInt(4);
+        float accuracy = atPosition.getInt(5);
+        double latitude = atPosition.getDouble(6);
+        double longitude = atPosition.getDouble(7);
+        return new WifiMeasurement(date,SSID,BSSID,dBm, accuracy, latitude,longitude);
+    }
+
+    public void printLastLine(){
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/managers/WifiMeasurementManager.java]: This is a debug method right? It should be removed before we merge - right?

File: app/src/main/java/de/tum/in/tumcampusapp/managers/WifiMeasurementManager.java [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+            for (int i = 0; i < c.getColumnCount(); i++) {
+                message += " " + c.getString(i);
+            }
+            if (!message.isEmpty()) {
+                Utils.log("WifiMeasurement" + message);
+            }
+        }
+    }
+
+    public void printToLog(){
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/managers/WifiMeasurementManager.java]: Also debug?

File: app/src/main/java/de/tum/in/tumcampusapp/managers/WifiMeasurementManager.java [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    /**
+     * This method iterates through all local database entries and tries to store them
+     * to the server. If an error during network transmission occurs, it tries a maximum of
+     * maxRetries times and waits waitTimeInMillis. It then purges the local database.
+     * @param maxRetries
+     * The amount of maximum retries for sending a single measurement
+     * @param waitTimeInMillis
+     * The wait time after a failed attempt to send a measurement to the server
+     */
+    public void uploadMeasurementsToRemote(int maxRetries, int waitTimeInMillis) {
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/managers/WifiMeasurementManager.java]: Oh gawd no - this is a nice ddos on our server. 
So for every measurements you are executing one HTTP request. Good idea, but really bad if 10k clients do this.

Please select all measurements and send an array of measurements to the backend.

File: app/src/main/java/de/tum/in/tumcampusapp/services/ScanResultsAvailableReceiver.java [hunk 12]
```diff
@@ -1,56 +1,83 @@
...
 import de.tum.in.tumcampusapp.managers.EduroamManager;
+import de.tum.in.tumcampusapp.models.tumcabe.WifiMeasurement;
 
 /**
  * Listens for android's ScanResultsAvailable broadcast and checks if eduroam is nearby.
  * If yes and eduroam has not been setup by now it shows an according notification.
  */
 public class ScanResultsAvailableReceiver extends BroadcastReceiver {
     private static final String SHOULD_SHOW = "wifi_setup_notification_dismissed";
+    private static final int MINIMUM_BATTERY_PERCENTAGE = 50;
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/services/ScanResultsAvailableReceiver.java]: Please describe these vars

File: app/src/main/java/de/tum/in/tumcampusapp/services/ScanResultsAvailableReceiver.java [hunk 13]
```diff
@@ -1,56 +1,83 @@
...
+     * The SyncManager then takes care of sending the Wifi measurements to the server in a given time
+     * interval.
+     */
     public void onReceive(Context context, Intent intent) {
         if (!intent.getAction().equals(WifiManager.SCAN_RESULTS_AVAILABLE_ACTION)) {
             return;
         }
-
-        // Test if user has eduroam configured already
-        if (EduroamManager.getEduroamConfig(context) != null || NetUtils.isConnected(context) || Build.VERSION.SDK_INT < 18) {
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/services/ScanResultsAvailableReceiver.java]: So you just removed the Eduroam config ehh? Funny idea. Not gonna happen.

File: app/src/main/java/de/tum/in/tumcampusapp/services/ScanResultsAvailableReceiver.java [hunk 14]
```diff
@@ -1,56 +1,83 @@
...
+        //schedule another scan
+        if (networkFound && Utils.getBatteryLevel(context) >= MINIMUM_BATTERY_PERCENTAGE){
+            wifiScanHandler.startRepetition();
+            Utils.log("WifiScanHandler rescheduled");
+        }else{
+            Utils.log("WifiScanHandler stopped");
+        }
+
+        // Test if user has eduroam configured already
+        if (EduroamManager.getEduroamConfig(context) != null || NetUtils.isConnected(context) || Build.VERSION.SDK_INT < 18) {
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/services/ScanResultsAvailableReceiver.java]: How does this make ANY sense down here?

File: app/src/main/java/de/tum/in/tumcampusapp/services/ScanResultsAvailableReceiver.java [hunk 15]
```diff
@@ -60,6 +87,31 @@ public void onReceive(Context context, Intent intent) {
...
+     * This method stores wifi scan results to the server. When they first get created by the
+     * ScanResultsAvailable's onReceive method, they lack gps information for creating a heatmap.
+     * Therefore we request an update from the WifiMeasurementLocationListener, passing the incomplete WifiMeasurement.
+     * The WifiMeasurementLocationListener then takes care of adding the location information, whenever it is ready.
+     * @param context
+     * @param scanResult
+     */
+
+    private void storeWifiMeasurement(Context context, ScanResult scanResult){
+        if (ActivityCompat.checkSelfPermission(context, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED){
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/services/ScanResultsAvailableReceiver.java]: Please check this permission when starting the service, not on every entry.

File: app/src/main/java/de/tum/in/tumcampusapp/services/ScanResultsAvailableReceiver.java [hunk 16]
```diff
@@ -60,6 +87,31 @@ public void onReceive(Context context, Intent intent) {
...
+            Utils.log("Permission not granted for FINE location scan in ScanResultsAvailableReceiver");
+            return;
+        }
+        LocationManager locationManager = (LocationManager) context.getApplicationContext().getSystemService(Context.LOCATION_SERVICE);
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/services/ScanResultsAvailableReceiver.java]: You should create the LocationManager once, and reuse it afterwards.

File: app/src/main/java/de/tum/in/tumcampusapp/services/SendWifiMeasurementService.java [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+import android.app.IntentService;
+import android.content.Intent;
+
+import de.tum.in.tumcampusapp.managers.WifiMeasurementManager;
+
+/**
+ * This service is getting used by StartSyncReceiver to send the accumulated WifiMeasurements
+ * every X hours
+ */
+public class SendWifiMeasurementService extends IntentService {
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/services/SendWifiMeasurementService.java]: Why do we need a service for this?

Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/services/SendWifiMeasurementService.java]: We already have other services that do regular updates with the backend

File: app/src/main/java/de/tum/in/tumcampusapp/services/StartSyncReceiver.java [hunk 18]
```diff
@@ -25,7 +26,12 @@ private static void setAlarm(Context context) {
         // Set alarm
         AlarmManager alarm = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
         alarm.cancel(pendingIntent);
-        alarm.set(AlarmManager.RTC, System.currentTimeMillis() + StartSyncReceiver.START_INTERVAL, pendingIntent);
+        if(Build.VERSION.SDK_INT < 19) {
+            alarm.set(AlarmManager.RTC, System.currentTimeMillis() + StartSyncReceiver.START_INTERVAL, pendingIntent);
+        }
+        else{
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/services/StartSyncReceiver.java]: iuehhhh, please put the else on the same line as your brackets.

**Comment by t2m4k1** [comment]
@kordianbruck 
"You're not authorized to merge this pull request." Help please sensai! The feature is disabled as we discussed and can be merged.

**Comment by kordianbruck** [comment]
@t2m4k1 yea, I deactivated that again, as I didn't want random code in the master branch :P
